File: apps/ios/BankCore/Networking/JWT.swift
import Foundation

enum JWT {
    struct Payload: Decodable {
        let sub: String
        let taxId: String?
        let name: String?
        let exp: TimeInterval?

        enum CodingKeys: String, CodingKey {
            case sub, name, exp
            case taxId = "tax_id"
        }

        var userId: UUID? { UUID(uuidString: sub) }

        var isExpired: Bool {
            guard let exp else { return false }
            return Date().timeIntervalSince1970 >= exp
        }
    }

    static func payload(from token: String) throws -> Payload {
        let parts = token.split(separator: ".")
        guard parts.count >= 2 else {
            throw APIError(status: 0, detail: "Token JWT inválido.")
        }
        var base64 = String(parts[1])
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let pad = (4 - base64.count % 4) % 4
        base64 += String(repeating: "=", count: pad)
        guard let data = Data(base64Encoded: base64) else {
            throw APIError(status: 0, detail: "Não foi possível ler o JWT.")
        }
        return try JSONDecoder().decode(Payload.self, from: data)
    }
}

File: apps/android/app/src/main/java/br/vortex/bankcore/data/Jwt.kt
package br.vortex.bankcore.data

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import android.util.Base64
import java.nio.charset.StandardCharsets

data class JwtPayload(
    val sub: String?,
    val taxId: String?,
    val name: String?,
    val exp: Long?,
) {
    val userId: String? get() = sub
    val isExpired: Boolean
        get() = exp != null && System.currentTimeMillis() / 1000 >= exp
}

object Jwt {
    fun payload(token: String): JwtPayload {
        val parts = token.split('.')
        if (parts.size < 2) throw ApiException(0, "Token JWT inválido.")
        var base64 = parts[1].replace('-', '+').replace('_', '/')
        val pad = (4 - base64.length % 4) % 4
        base64 += "=".repeat(pad)
        val json = try {
            String(Base64.decode(base64, Base64.DEFAULT), StandardCharsets.UTF_8)
        } catch (_: Exception) {
            throw ApiException(0, "Não foi possível ler o JWT.")
        }
        val obj = Json.parseToJsonElement(json).jsonObject
        return JwtPayload(
            sub = obj["sub"]?.jsonPrimitive?.contentOrNull,
            taxId = obj["tax_id"]?.jsonPrimitive?.contentOrNull,
            name = obj["name"]?.jsonPrimitive?.contentOrNull,
            exp = obj["exp"]?.jsonPrimitive?.longOrNull,
        )
    }
}
